### DISCUSSION TITLE
ilogtail  input skywalking v3插件和flusher   flusher_grpc插件同时使用程序panic

### DISCUSSION THREAD

**Discussion by yanggelinux** [alibaba/ilogtail#563] [Help] [answered]
我用ilogtail 的skywalking v3 插件收集trace 数据，只用flusher_stdout 能正常输出到控制台，我想用flusher_grpc 把收集的数据发走，添加flusher_grpc ilogtail就会panic，我把这个配置文件中的flusher_grpc注释掉后，只要user_yaml_config.d目录下别的配置文件有使用了flusher_grpc 整个ilogtail 也会panic。
下面是报错信息
panic: protobuf tag not enough fields in JVMMetricCollection.state: 

goroutine 61 [running]:
github.com/gogo/protobuf/proto.(*unmarshalInfo).computeUnmarshalInfo(0xc0004e2fa0)
  /root/go/pkg/mod/github.com/gogo/protobuf@v1.3.2/proto/table_unmarshal.go:341 +0x138a
github.com/gogo/protobuf/proto.(*unmarshalInfo).unmarshal(0xc0004e2fa0, {0x7fa6b8d3cc40?}, {0xc000ca6000, 0x2b8a, 0x2b8a})
  /root/go/pkg/mod/github.com/gogo/protobuf@v1.3.2/proto/table_unmarshal.go:138 +0x67
github.com/gogo/protobuf/proto.(*InternalMessageInfo).Unmarshal(0xc000091b00?, {0x7fa68ac8cba8, 0xc00034bc20}, {0xc000ca6000?, 0x2b8a?, 0x2b8a?})
  /root/go/pkg/mod/github.com/gogo/protobuf@v1.3.2/proto/table_unmarshal.go:63 +0xd0
github.com/gogo/protobuf/proto.(*Buffer).Unmarshal(0xc000cb38c8, {0x7fa68ac8cba8, 0xc00034bc20})
  /root/go/pkg/mod/github.com/gogo/protobuf@v1.3.2/proto/decode.go:424 +0x153
github.com/gogo/protobuf/proto.Unmarshal({0xc000ca6000, 0x2b8a, 0x2b8a}, {0x7fa68ac8cba8, 0xc00034bc20})
  /root/go/pkg/mod/github.com/gogo/protobuf@v1.3.2/proto/decode.go:342 +0xe6
github.com/alibaba/ilogtail/pkg/protocol.Codec.Unmarshal({}, {0xc000ca6000, 0x2b8a, 0x2b8a}, {0x7fa6b8cdcf80, 0xc00034bc20})
  /src/pkg/protocol/sls_logs.pb.helper.go:38 +0x65
google.golang.org/grpc.(*Server).processUnaryRPC.func2({0x7fa6b8cdcf80?, 0xc00034bc20})
  /root/go/pkg/mod/google.golang.org/grpc@v1.49.0/server.go:1278 +0x224
github.com/alibaba/ilogtail/plugins/input/skywalkingv3/skywalking/network/language/agent/v3._JVMMetricReportService_Collect_Handler({0x7fa6b8af1c20?, 0xc0003533e0}, {0x7fa6b8f25500, 0xc0004d3350}, 0xc0001979d0, 0x0)
  /src/plugins/input/skywalkingv3/skywalking/network/language/agent/v3/JVMMetric_grpc.pb.go:70 +0x63
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000000960, {0x7fa6b8f32938, 0xc000a1b380}, 0xc0001c9440, 0xc000353470, 0x7fa6ba3eb420, 0x0)
  /root/go/pkg/mod/google.golang.org/grpc@v1.49.0/server.go:1301 +0xb0b
google.golang.org/grpc.(*Server).handleStream(0xc000000960, {0x7fa6b8f32938, 0xc000a1b380}, 0xc0001c9440, 0x0)
  /root/go/pkg/mod/google.golang.org/grpc@v1.49.0/server.go:1642 +0xa1b
google.golang.org/grpc.(*Server).serveStreams.func1.2()
  /root/go/pkg/mod/google.golang.org/grpc@v1.49.0/server.go:938 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
  /root/go/pkg/mod/google.golang.org/grpc@v1.49.0/server.go:936 +0x28a

**Comment by yanggelinux** [comment]
func (Codec) Unmarshal(data []byte, v interface{}) error {
	vv, ok := v.(proto.Message)
	if !ok {
		return fmt.Errorf("failed to unmarshal, message is %T, want proto.Message", v)
	}
	return proto.Unmarshal(data, vv)
}

/src/pkg/protocol/sls_logs.pb.helper.go:38 

看了一下是这里的代码报错，猜测sls_logs 是proto2,skywalking 是proto3，版本不一致导致。

**Reply by qiansheng91** [reply]
skywalking v3插件会把skywalking数据转换成sls log的格式，GRPC Server的服务是什么服务？

**Reply by yanggelinux** [reply]
server 服务是go语言的grpc服务，我直接把ilogtail sls log pb文件拷贝到server 端使用，本来收集log都能正常收集，想收集trace数据 添加 skywalking 插件后，程序开始panic。

**Reply by yanggelinux** [reply]
encoding.RegisterCodec(new(protocol.Codec))  我把flusher_grpc  Init 方法中的这行代码注释掉，程序不panic了，但是数据发送到server 失败，ilogtail日志 开始报AlarmType:FLUSH_DATA_ALARM	flush data error:	:rpc error: code = Internal desc = grpc: error while marshaling: proto: required field sls_logs.Log.Content.Value not set	的错误。

**Reply by yanggelinux** [reply]
看了一下还是sls_logs.pb.helper.go 文件中的proto.Unmarshal 报错导致程序panic

**Reply by yanggelinux** [reply]
> skywalking v3插件会把skywalking数据转换成sls log的格式，GRPC Server的服务是什么服务？

server 服务是go语言的grpc服务，我直接把ilogtail sls log pb文件拷贝到server 端使用，本来收集log都能正常收集，想收集trace数据 添加 skywalking 插件后，程序开始panic。

**Reply by yanggelinux** [reply]
我把  sls_logs.pb.helper.go 文件中 "github.com/gogo/protobuf/proto" 替换成"github.com/golang/protobuf/proto" 程序不panic了，但是grpc 发送数据还是失败 还是报 ilogtail日志 开始报AlarmType:FLUSH_DATA_ALARM flush data error: :rpc error: code = Internal desc = grpc: error while marshaling: proto: required field sls_logs.Log.Content.Value not set 的错误。

**Comment by yanggelinux** [comment]
我把 sls_logs.pb.helper.go 文件中 "github.com/gogo/protobuf/proto" 替换成"github.com/golang/protobuf/proto" 程序不panic了，但是grpc 发送数据还是失败 还是报 ilogtail日志 开始报AlarmType:FLUSH_DATA_ALARM flush data error: :rpc error: code = Internal desc = grpc: error while marshaling: proto: required field sls_logs.Log.Content.Value not set 的错误。

**Reply by yanggelinux** [reply]
换了proto 报后log能正常发送，trace 还是不行，但是程序不panic了，也不会因为比的配置文件有flasher_grpc 程序panic

**Reply by yanggelinux** [reply, marked as answer]
required field sls_logs.Log.Content.Value not set 这个报错是因为skywalking trace 数据有空值导致，处理了空值就不报错了
